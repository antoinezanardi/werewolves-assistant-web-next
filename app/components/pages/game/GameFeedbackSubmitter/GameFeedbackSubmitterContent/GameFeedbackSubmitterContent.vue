<template>
  <div
    id="game-feedback-submitter-content"
  >
    <form class="flex flex-col gap-2">
      <div class="flex flex-col gap-4 justify-around md:flex-row">
        <GameFeedbackSubmitterRating
          id="game-feedback-submitter-rating"
          v-model="createGameFeedbackDto.score"
          class="w-auto"
        />

        <GameFeedbackSubmitterEncounteredError
          id="game-feedback-submitter-encountered-error"
          v-model="createGameFeedbackDto.hasEncounteredError"
          class="w-auto"
        />
      </div>

      <PrimeVueDivider/>

      <GameFeedbackSubmitterReview
        id="game-feedback-submitter-review"
        v-model="createGameFeedbackDto.review"
      />
    </form>
  </div>
</template>

<script setup lang="ts">
import GameFeedbackSubmitterEncounteredError from "~/components/pages/game/GameFeedbackSubmitter/GameFeedbackSubmitterContent/GameFeedbackSubmitterEncounteredError/GameFeedbackSubmitterEncounteredError.vue";
import GameFeedbackSubmitterRating from "~/components/pages/game/GameFeedbackSubmitter/GameFeedbackSubmitterContent/GameFeedbackSubmitterRating/GameFeedbackSubmitterRating.vue";
import GameFeedbackSubmitterReview from "~/components/pages/game/GameFeedbackSubmitter/GameFeedbackSubmitterContent/GameFeedbackSubmitterReview/GameFeedbackSubmitterReview.vue";
import type { CreateGameFeedbackDto } from "~/composables/api/game/dto/create-game-feedback/create-game-feedback.dto";

const createGameFeedbackDto = defineModel<CreateGameFeedbackDto>({ required: true });
</script>