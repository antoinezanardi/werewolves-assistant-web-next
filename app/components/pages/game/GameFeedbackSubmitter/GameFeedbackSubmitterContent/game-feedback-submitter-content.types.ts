import type { CreateGameFeedbackDto } from "~/composables/api/game/dto/create-game-feedback/create-game-feedback.dto";

type GameFeedbackSubmitterContentProps = {
  modelValue: CreateGameFeedbackDto;
};

export type {
  GameFeedbackSubmitterContentProps,
};