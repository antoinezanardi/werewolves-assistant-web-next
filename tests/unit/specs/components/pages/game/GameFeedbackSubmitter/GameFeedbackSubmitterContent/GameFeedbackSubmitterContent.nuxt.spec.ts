import type { Ref } from "vue";
import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import type { VueVm } from "@tests/unit/utils/types/vue-test-utils.types";
import { createFakeCreateGameFeedbackDto } from "@tests/unit/utils/factories/composables/api/game/dto/create-game-feedback/create-game-feedback.dto.factory";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { GameFeedbackSubmitterContentProps } from "~/components/pages/game/GameFeedbackSubmitter/GameFeedbackSubmitterContent/game-feedback-submitter-content.types";
import GameFeedbackSubmitterContent from "~/components/pages/game/GameFeedbackSubmitter/GameFeedbackSubmitterContent/GameFeedbackSubmitterContent.vue";
import type GameFeedbackSubmitterRating from "~/components/pages/game/GameFeedbackSubmitter/GameFeedbackSubmitterContent/GameFeedbackSubmitterRating/GameFeedbackSubmitterRating.vue";
import type GameFeedbackSubmitterReview from "~/components/pages/game/GameFeedbackSubmitter/GameFeedbackSubmitterContent/GameFeedbackSubmitterReview/GameFeedbackSubmitterReview.vue";
import type { CreateGameFeedbackDto } from "~/composables/api/game/dto/create-game-feedback/create-game-feedback.dto";

describe("Game Feedback Submitter Content Component", () => {
  const defaultCreateGameFeedbackDto = createFakeCreateGameFeedbackDto({
    score: 2,
    review: "This game is awesome!",
  });
  const defaultProps: GameFeedbackSubmitterContentProps = {
    modelValue: defaultCreateGameFeedbackDto,
  };
  let wrapper: ReturnType<typeof mount<typeof GameFeedbackSubmitterContent>>;

  async function mountGameFeedbackSubmitterContentComponent(options: ComponentMountingOptions<typeof GameFeedbackSubmitterContent> = {}):
  Promise<ReturnType<typeof mount<typeof GameFeedbackSubmitterContent>>> {
    return mountSuspendedComponent(GameFeedbackSubmitterContent, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameFeedbackSubmitterContentComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Content", () => {
    describe("Rating", () => {
      it("should pass create create game feedback score to rating dto when rendered.", () => {
        const rating = wrapper.findComponent<typeof GameFeedbackSubmitterRating>("#game-feedback-submitter-rating");

        expect(rating.props("modelValue")).toBe(defaultProps.modelValue.score);
      });

      it("should change create game feedback score when update model value event is emitted.", async() => {
        const rating = wrapper.findComponent<typeof GameFeedbackSubmitterRating>("#game-feedback-submitter-rating");
        const newModelValue = 3;
        (rating.vm as VueVm).$emit("update:modelValue", newModelValue);
        await nextTick();

        expect((wrapper.vm as unknown as { createGameFeedbackDto: Ref<CreateGameFeedbackDto> }).createGameFeedbackDto.value.score).toBe(newModelValue);
      });
    });

    describe("Review", () => {
      it("should pass create create game feedback review to review dto when rendered.", () => {
        const review = wrapper.findComponent<typeof GameFeedbackSubmitterReview>("#game-feedback-submitter-review");

        expect(review.props("modelValue")).toBe(defaultProps.modelValue.review);
      });

      it("should change create game feedback review when update model value event is emitted.", async() => {
        const review = wrapper.findComponent<typeof GameFeedbackSubmitterReview>("#game-feedback-submitter-review");
        const newModelValue = "This game is not awesome!";
        (review.vm as VueVm).$emit("update:modelValue", newModelValue);
        await nextTick();

        expect((wrapper.vm as unknown as { createGameFeedbackDto: Ref<CreateGameFeedbackDto> }).createGameFeedbackDto.value.review).toBe(newModelValue);
      });
    });
  });
});