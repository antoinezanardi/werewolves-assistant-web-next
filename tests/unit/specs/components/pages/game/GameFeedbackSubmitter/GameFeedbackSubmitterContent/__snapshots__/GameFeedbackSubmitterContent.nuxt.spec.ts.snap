// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Feedback Submitter Content Component > should match snapshot when rendered. 1`] = `
"<div id="game-feedback-submitter-content">
  <form class="flex flex-col gap-2">
    <div class="flex flex-col gap-4 justify-around md:flex-row">
      <game-feedback-submitter-rating-stub modelvalue="2" id="game-feedback-submitter-rating" class="w-auto"></game-feedback-submitter-rating-stub>
      <game-feedback-submitter-encountered-error-stub modelvalue="false" id="game-feedback-submitter-encountered-error" class="w-auto"></game-feedback-submitter-encountered-error-stub>
    </div>
    <divider-stub></divider-stub>
    <game-feedback-submitter-review-stub modelvalue="This game is awesome!" id="game-feedback-submitter-review"></game-feedback-submitter-review-stub>
  </form>
</div>"
`;
